### PR TITLE
[WIP]Vae preprocessor refactor (PR1)

### PR DESCRIPTION
**Pull request by yiyixuxu**
VaeImageProcessor.preprocess refactor

* refactored `VaeImageProcessor` a little bit
  * allow passing optional `height` and `width` argument to `resize()`
  * add `convert_to_rgb`
* refactored `prepare_latents` method for img2img pipelines so that if we pass latents directly as image input, it will not encode it again 
* added a test in `test_pipelines_common.py` to test latents as image inputs
* refactored __img2img__ pipelines that __accept latents__ as `image`: controlnet img2img, stable diffusion img2img , instruct_pix2pix

### CODE REVIEW

**Review by patrickvonplaten — COMMENTED** [review]
Nice start! Can we maybe first merge the PR that adds VAE preprocess and then merge this one? Otherwise people will see lots of deprecation warnings :sweat_smile:

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py [hunk 1]
```diff
@@ -915,14 +899,12 @@ def __call__(
             (nsfw) content, according to the `safety_checker`.
         """
         # 0. Default height and width to unet
-        height, width = self._default_height_width(height, width, image)
+        # height, width = self._default_height_width(height, width, image)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py:902]: I think we don't need this `default_height_width` function anymore because the `image_processor.preprocess` will resize the image to default size

File: src/diffusers/image_processor.py [hunk 2]
```diff
@@ -146,7 +181,12 @@ def preprocess(
 
         elif isinstance(image[0], torch.Tensor):
             image = torch.cat(image, axis=0) if image[0].ndim == 4 else torch.stack(image, axis=0)
-            _, _, height, width = image.shape
+            _, channel, height, width = image.shape
+
+            # don't need any preprocess if the image is latents
+            if channel == 4:
```
Comment (patrickvonplaten) [inline src/diffusers/image_processor.py:187]: nice! Good for me!

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py [hunk 3]
```diff
@@ -624,38 +626,14 @@ def check_image(self, image, prompt, prompt_embeds):
...
-                image = images
-
-                image = np.concatenate(image, axis=0)
-                image = np.array(image).astype(np.float32) / 255.0
-                image = image.transpose(0, 3, 1, 2)
-                image = torch.from_numpy(image)
-            elif isinstance(image[0], torch.Tensor):
-                image = torch.cat(image, dim=0)
-
+        image = self.control_image_processor.preprocess(image).to(dtype=torch.float32)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:636]:
```suggestion
        image = self.control_image_processor.preprocess(image,  width=width, height=height).to(dtype=torch.float32)
```

shouldn't we pass width and height here?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py [hunk 4]
```diff
@@ -198,7 +197,10 @@ def __init__(
             feature_extractor=feature_extractor,
         )
         self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
-        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor, do_convert_rgb=True)
+        self.control_image_processor = VaeImageProcessor(
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py:201]: nice!

**Review by patrickvonplaten — APPROVED** [review]
This looks very nice to me! 

Think my only feedback is this: https://github.com/huggingface/diffusers/pull/3557/files#r1211746978

We should pass the height and width here as well no since it was possible before?

**Review by pcuenca — COMMENTED** [review]
Awesome!

~~I think there are some spurious `width` and `height` in the ControlNet pipeline.~~

Sorry, I reviewed an intermediate version that was inconsistent.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py [hunk 5]
```diff
@@ -633,29 +635,7 @@ def prepare_image(
...
-                image = images
-
-                image = np.concatenate(image, axis=0)
-                image = np.array(image).astype(np.float32) / 255.0
-                image = image.transpose(0, 3, 1, 2)
-                image = torch.from_numpy(image)
-            elif isinstance(image[0], torch.Tensor):
-                image = torch.cat(image, dim=0)
-
+        image = self.control_image_processor.preprocess(image, width=width, height=height).to(dtype=torch.float32)
```
Comment (pcuenca) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:638]: I think the order is `height` then `width`:

```suggestion
        image = self.control_image_processor.preprocess(image, height=height, width=width).to(dtype=torch.float32)
```

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py [hunk 6]
```diff
@@ -660,29 +663,7 @@ def prepare_control_image(
...
-                image = images
-
-                image = np.concatenate(image, axis=0)
-                image = np.array(image).astype(np.float32) / 255.0
-                image = image.transpose(0, 3, 1, 2)
-                image = torch.from_numpy(image)
-            elif isinstance(image[0], torch.Tensor):
-                image = torch.cat(image, dim=0)
-
+        image = self.control_image_processor.preprocess(image, width=width, height=height).to(dtype=torch.float32)
```
Comment (pcuenca) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_img2img.py:666]:
```suggestion
        image = self.control_image_processor.preprocess(image, height=height, width=width).to(dtype=torch.float32)
```

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py [hunk 7]
```diff
@@ -787,29 +795,7 @@ def prepare_control_image(
...
-                image = images
-
-                image = np.concatenate(image, axis=0)
-                image = np.array(image).astype(np.float32) / 255.0
-                image = image.transpose(0, 3, 1, 2)
-                image = torch.from_numpy(image)
-            elif isinstance(image[0], torch.Tensor):
-                image = torch.cat(image, dim=0)
-
+        image = self.control_image_processor.preprocess(image, width=width, height=height).to(dtype=torch.float32)
```
Comment (pcuenca) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py:798]:
```suggestion
        image = self.control_image_processor.preprocess(image, height=height, width=width).to(dtype=torch.float32)
```

**Review by pcuenca — APPROVED** [review]

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py [hunk 8]
```diff
@@ -723,25 +733,24 @@ def __call__(
...
-            image = self.decode_latents(latents)
+            image = latents
             has_nsfw_concept = None
 
+        if has_nsfw_concept is None:
+            do_denormalize = [True] * image.shape[0]
+        else:
+            do_denormalize = [not has_nsfw for has_nsfw in has_nsfw_concept]
+
+        image = self.image_processor.postprocess(image, output_type=output_type, do_denormalize=do_denormalize)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:748]: @patrickvonplaten I refactored the 4x upscaler here (just `preprocess` and `postprocess`, not accepting latents) 
however I think I changed the logic of `postprocess` here, i.e. if `output_type ==pt`, currently it will return a pytorch tensor that's unnormalized, which is inconsistent with `image_processor.postprocess`. Let me know if we actually intend to return pytorch tensor between `[-1,1]` for this pipeline though

File: src/diffusers/image_processor.py [hunk 9]
```diff
@@ -49,6 +49,7 @@ def __init__(
         vae_scale_factor: int = 8,
         resample: str = "lanczos",
         do_normalize: bool = True,
+        do_convert_rgb: bool = False,
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/image_processor.py:55]: +1

File: src/diffusers/image_processor.py [hunk 10]
```diff
@@ -30,14 +30,17 @@ class VaeImageProcessor(ConfigMixin):
...
+            `height` and `width` arguments from `preprocess` method
         vae_scale_factor (`int`, *optional*, defaults to `8`):
             VAE scale factor. If `do_resize` is True, the image will be automatically resized to multiples of this
             factor.
         resample (`str`, *optional*, defaults to `lanczos`):
             Resampling filter to use when resizing the image.
         do_normalize (`bool`, *optional*, defaults to `True`):
             Whether to normalize the image to [-1,1]
+        do_convert_rgb (`bool`, *optional*, defaults to be `False`):
+            Whether to converts the images to RGB format.
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:43]:
```suggestion
            Whether to convert the images to RGB format.
```

File: src/diffusers/image_processor.py [hunk 11]
```diff
@@ -68,6 +72,18 @@ def numpy_to_pil(images):
 
         return pil_images
 
+    @staticmethod
+    def pil_to_numpy(images):
+        """
+        Convert a PIL image or a list of PIL images to numpy image
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:78]:
```suggestion
        Convert a PIL image or a list of PIL images to numpy arrays.
```

Can we also have type annotations here?

File: src/diffusers/image_processor.py [hunk 12]
```diff
@@ -68,6 +72,18 @@ def numpy_to_pil(images):
...
+        """
+        if not isinstance(images, list):
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:80]: Is there a need to also check if `image` is of type `PIL.Image.Image`?

File: src/diffusers/image_processor.py [hunk 13]
```diff
@@ -101,18 +117,39 @@ def denormalize(images):
...
-        return images
+        if height is None:
+            height = image.height
+        if width is None:
+            width = image.width
+
+        width, height = (
+            x - x % self.config.vae_scale_factor for x in (width, height)
+        )  # resize to integer multiple of vae_scale_factor
+        image = image.resize((width, height), resample=PIL_INTERPOLATION[self.config.resample])
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:145]: 👌

File: src/diffusers/image_processor.py [hunk 14]
```diff
@@ -126,10 +163,11 @@ def preprocess(
...
         if isinstance(image[0], PIL.Image.Image):
+            if self.config.do_convert_rgb:
+                image = [self.convert_to_rgb(i) for i in image]
             if self.config.do_resize:
-                image = [self.resize(i) for i in image]
-            image = [np.array(i).astype(np.float32) / 255.0 for i in image]
-            image = np.stack(image, axis=0)  # to np
+                image = [self.resize(i, height, width) for i in image]
+            image = self.pil_to_numpy(image)  # to np
             image = self.numpy_to_pt(image)  # to pt
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:171]: Why do we simultaneously do this?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py [hunk 15]
```diff
@@ -172,7 +171,10 @@ def __init__(
             feature_extractor=feature_extractor,
         )
         self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
-        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor, do_convert_rgb=True)
+        self.control_image_processor = VaeImageProcessor(
+            vae_scale_factor=self.vae_scale_factor, do_convert_rgb=True, do_normalize=False
+        )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:177]: Nice to see this coming to fruition.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py [hunk 16]
```diff
@@ -592,21 +589,26 @@ def check_inputs(
...
 
-        if not image_is_pil and not image_is_tensor and not image_is_pil_list and not image_is_tensor_list:
+        if (
+            not image_is_pil
+            and not image_is_tensor
+            and not image_is_np
+            and not image_is_pil_list
+            and not image_is_tensor_list
+            and not image_is_np_list
+        ):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:604]: That's a lot of conditions hahaha.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py [hunk 17]
```diff
@@ -592,21 +589,26 @@ def check_inputs(
...
             )
 
         if image_is_pil:
             image_batch_size = 1
-        elif image_is_tensor:
-            image_batch_size = image.shape[0]
-        elif image_is_pil_list:
-            image_batch_size = len(image)
-        elif image_is_tensor_list:
+        else:
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py:611]: Yes sir.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py [hunk 18]
```diff
@@ -633,29 +635,7 @@ def prepare_image(
...
-                image = images
-
-                image = np.concatenate(image, axis=0)
-                image = np.array(image).astype(np.float32) / 255.0
-                image = image.transpose(0, 3, 1, 2)
-                image = torch.from_numpy(image)
-            elif isinstance(image[0], torch.Tensor):
-                image = torch.cat(image, dim=0)
-
+        image = self.control_image_processor.preprocess(image, height=height, width=width).to(dtype=torch.float32)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:638]: So, all of this logic is handled by `preprocess()` now? That's amazing!

File: src/diffusers/pipelines/unidiffuser/pipeline_unidiffuser.py [hunk 19]
```diff
@@ -34,6 +35,11 @@
 
 # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.preprocess
 def preprocess(image):
+    warnings.warn(
+        "The preprocess method is deprecated and will be removed in a future version. Please"
+        " use VaeImageProcessor.preprocess instead",
+        FutureWarning,
+    )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/unidiffuser/pipeline_unidiffuser.py:42]: So, our plan is to refactor this in a future PR, yes?

**Review by sayakpaul — APPROVED** [review]
Awesome!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py (hunk 8 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:748]: If previously, the latent upscaler was returning unnormalized tensors, I would prefer keeping it that way to avoid any unforeseen consequences? 

Maybe, we could add a flag to `image_processor.postprocess` to check if normalization is needed? To me, that is a cleaner and more idiomatic approach.

File: src/diffusers/pipelines/unidiffuser/pipeline_unidiffuser.py (hunk 19 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/unidiffuser/pipeline_unidiffuser.py:42]: No this function should be fully deprecated and removed in the future

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py [hunk 20]
```diff
@@ -94,7 +94,7 @@ def __init__(
             scheduler=scheduler,
         )
         self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
-        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.image_processor = VaeImageProcessor(vae_scale_factor=64, resample="bicubic")
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:97]: Hmmm why do we do this change here?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:748]: @yiyixuxu could we for now make sure that the output stays exactly the same. I.e. we should not change the behavior of the pipelines in any way IMO.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_instruction_pix2pix.py [hunk 21]
```diff
@@ -199,6 +204,28 @@ def test_stable_diffusion_pix2pix_euler(self):
     def test_inference_batch_single_identical(self):
         super().test_inference_batch_single_identical(expected_max_diff=3e-3)
 
+    # Overwrite the default test_latents_inputs because pix2pix encode the image differently
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_instruction_pix2pix.py:207]: Nice!

**Review by patrickvonplaten — COMMENTED** [review]
I think we can merge this more or less. The final missing piece seems to be this: https://github.com/huggingface/diffusers/pull/3557/files#r1214586673

Can we make sure that we don't change the output behavior in any way?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py (hunk 20 above)
Comment (yiyixuxu) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:97]: just to match how the  `preprocess` method was implemented for this pipeline https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py#L46
it use default resample ("bicubic") vs the img2img preprocess use "lanczos" https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L85
I didn't make the change before because we were just using the VaeImageProcessor for `postprocess`

**Comment by yiyixuxu** [comment]
@patrickvonplaten 
reverted changes I made to x4 upscaler and created a separate issue here https://github.com/huggingface/diffusers/issues/3654

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py (hunk 20 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:97]: I think this is not correct however. The `vae_scale_factor` is not 64 here but should be `self.vae_scale_factor` which corresponds to 8. The 64 was just used for images that had sizes that are **not** multiple of 64, but we can change this to 8.
```suggestion
        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor, resample="bicubic")
```

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:97]: Since this PR: https://github.com/huggingface/diffusers/pull/505 all stable diffusion models can work with images just being multiples of 8
